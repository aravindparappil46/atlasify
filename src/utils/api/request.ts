import axios, { type AxiosPromise, type Method } from 'axios';
import type { Link, Token } from '../../types';

export function apiRequestAuth(
  url: Link,
  method: Method,
  username: string,
  token: Token,
  data = {},
): AxiosPromise | null {
  axios.defaults.headers.common.Accept = 'application/json';
  axios.defaults.headers.common.Authorization = `Basic ${btoa(`${username}:${token}`)}`;
  axios.defaults.headers.common['Cache-Control'] = 'no-cache';
  axios.defaults.headers.common['Content-Type'] = 'application/json';
  return axios({ method, url, data });
}
