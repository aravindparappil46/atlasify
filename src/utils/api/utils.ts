import { Constants } from '../constants';

export function getAPIUrl(): URL {
  return new URL(Constants.ATLASSIAN_API);
}
