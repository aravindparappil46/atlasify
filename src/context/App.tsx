import { ipcRenderer, webFrame } from 'electron';
import {
  type ReactNode,
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useInterval } from '../hooks/useInterval';
import { useNotifications } from '../hooks/useNotifications';
import {
  type Account,
  type AccountNotifications,
  type AtlasifyError,
  type AtlasifyNotification,
  type AuthState,
  GroupBy,
  OpenPreference,
  type SettingsState,
  type SettingsValue,
  type Status,
  Theme,
} from '../types';
// import type { Category, ReadState } from '../utils/api/types';
import type { LoginAPITokenOptions } from '../utils/auth/types';
import {
  addAccount,
  hasAccounts,
  refreshAccount,
  removeAccount,
} from '../utils/auth/utils';
import {
  setAlternateIdleIcon,
  setAutoLaunch,
  setKeyboardShortcut,
  updateTrayTitle,
} from '../utils/comms';
import { Constants } from '../utils/constants';
import { getNotificationCount } from '../utils/notifications';
import { clearState, loadState, saveState } from '../utils/storage';
import { setTheme } from '../utils/theme';
import { zoomPercentageToLevel } from '../utils/zoom';

export const defaultAuth: AuthState = {
  accounts: [],
};

const defaultAppearanceSettings = {
  theme: Theme.SYSTEM,
  zoomPercentage: 100,
};

const defaultNotificationSettings = {
  groupBy: GroupBy.DATE,
  markAsReadOnOpen: true,
  delayNotificationState: false,
};

const defaultSystemSettings = {
  openLinks: OpenPreference.FOREGROUND,
  keyboardShortcut: true,
  showNotificationsCountInTray: true,
  showSystemNotifications: true,
  playSound: true,
  useAlternateIdleIcon: false,
  openAtStartup: false,
};

export const defaultFilters = {
  filterCategories: [],
  filterReadStates: [],
  filterProducts: [],
};

export const defaultSettings: SettingsState = {
  ...defaultAppearanceSettings,
  ...defaultNotificationSettings,
  ...defaultSystemSettings,
  ...defaultFilters,
};

interface AppContextState {
  auth: AuthState;
  isLoggedIn: boolean;
  loginWithAPIToken: (data: LoginAPITokenOptions) => void;
  logoutFromAccount: (account: Account) => void;

  status: Status;
  globalError: AtlasifyError;

  notifications: AccountNotifications[];
  fetchNotifications: () => Promise<void>;
  removeAccountNotifications: (account: Account) => Promise<void>;

  markNotificationRead: (notification: AtlasifyNotification) => Promise<void>;
  markNotificationUnread: (notification: AtlasifyNotification) => Promise<void>;

  markProductNotificationsRead: (
    notification: AtlasifyNotification,
  ) => Promise<void>;
  markProductNotificationsUnread: (
    notification: AtlasifyNotification,
  ) => Promise<void>;

  settings: SettingsState;
  clearFilters: () => void;
  resetSettings: () => void;
  updateSetting: (name: keyof SettingsState, value: SettingsValue) => void;
}

export const AppContext = createContext<Partial<AppContextState>>({});

export const AppProvider = ({ children }: { children: ReactNode }) => {
  const [auth, setAuth] = useState<AuthState>(defaultAuth);
  const [settings, setSettings] = useState<SettingsState>(defaultSettings);
  const {
    notifications,
    fetchNotifications,
    removeAccountNotifications,
    status,
    globalError,
    markNotificationRead,
    markNotificationUnread,
    markProductNotificationsRead,
    markProductNotificationsUnread,
  } = useNotifications();
  getNotificationCount;
  useEffect(() => {
    restoreSettings();
  }, []);

  useEffect(() => {
    setTheme(settings.theme);
  }, [settings.theme]);

  // biome-ignore lint/correctness/useExhaustiveDependencies: We only want fetchNotifications to be called for account changes
  useEffect(() => {
    fetchNotifications({ auth, settings });
  }, [auth.accounts]);

  useInterval(() => {
    fetchNotifications({ auth, settings });
  }, Constants.FETCH_NOTIFICATIONS_INTERVAL);

  useInterval(() => {
    for (const account of auth.accounts) {
      refreshAccount(account);
    }
  }, Constants.REFRESH_ACCOUNTS_INTERVAL);

  useEffect(() => {
    const count = getNotificationCount(notifications);

    if (settings.showNotificationsCountInTray && count > 0) {
      updateTrayTitle(count.toString());
    } else {
      updateTrayTitle();
    }
  }, [settings.showNotificationsCountInTray, notifications]);

  useEffect(() => {
    setKeyboardShortcut(settings.keyboardShortcut);
  }, [settings.keyboardShortcut]);

  useEffect(() => {
    ipcRenderer.on('atlasify:reset-app', () => {
      clearState();
      setAuth(defaultAuth);
      setSettings(defaultSettings);
    });
  }, []);

  const clearFilters = useCallback(() => {
    const newSettings = { ...settings, ...defaultFilters };
    setSettings(newSettings);
    saveState({ auth, settings: newSettings });
  }, [auth, settings]);

  const resetSettings = useCallback(() => {
    setSettings(defaultSettings);
    saveState({ auth, settings: defaultSettings });
  }, [auth]);

  const updateSetting = useCallback(
    (name: keyof SettingsState, value: SettingsValue) => {
      if (name === 'openAtStartup') {
        setAutoLaunch(value as boolean);
      }
      if (name === 'useAlternateIdleIcon') {
        setAlternateIdleIcon(value as boolean);
      }

      const newSettings = { ...settings, [name]: value };
      setSettings(newSettings);
      saveState({ auth, settings: newSettings });
    },
    [auth, settings],
  );

  const isLoggedIn = useMemo(() => {
    return hasAccounts(auth);
  }, [auth]);

  const loginWithAPIToken = useCallback(
    async ({ username, token }: LoginAPITokenOptions) => {
      // Should check if the token is valid

      const updatedAuth = await addAccount(auth, username, token);

      // TODO - fetch and set user data

      setAuth(updatedAuth);
      saveState({ auth: updatedAuth, settings });
    },
    [auth, settings],
  );

  const logoutFromAccount = useCallback(
    async (account: Account) => {
      // Remove notifications for account
      removeAccountNotifications(account);

      // Remove from auth state
      const updatedAuth = removeAccount(auth, account);
      setAuth(updatedAuth);
      saveState({ auth: updatedAuth, settings });
    },
    [auth, settings],
  );

  const restoreSettings = useCallback(async () => {
    const existing = loadState();

    // Restore settings before accounts to ensure filters are available before fetching notifications
    if (existing.settings) {
      setKeyboardShortcut(existing.settings.keyboardShortcut);
      setAlternateIdleIcon(existing.settings.useAlternateIdleIcon);
      setSettings({ ...defaultSettings, ...existing.settings });
      webFrame.setZoomLevel(
        zoomPercentageToLevel(existing.settings.zoomPercentage),
      );
    }

    if (existing.auth) {
      setAuth({ ...defaultAuth, ...existing.auth });

      // Refresh account data on app start
      for (const account of existing.auth.accounts) {
        await refreshAccount(account);
      }
    }
  }, []);

  const fetchNotificationsWithAccounts = useCallback(
    async () => await fetchNotifications({ auth, settings }),
    [auth, settings, fetchNotifications],
  );

  const markNotificationReadWithAccounts = useCallback(
    async (notification: AtlasifyNotification) =>
      await markNotificationRead({ auth, settings }, notification),
    [auth, settings, markNotificationRead],
  );

  const markNotificationUnreadWithAccounts = useCallback(
    async (notification: AtlasifyNotification) =>
      await markNotificationUnread({ auth, settings }, notification),
    [auth, settings, markNotificationUnread],
  );

  const markProductNotificationsReadWithAccounts = useCallback(
    async (notification: AtlasifyNotification) =>
      await markProductNotificationsRead({ auth, settings }, notification),
    [auth, settings, markProductNotificationsRead],
  );

  const markProductNotificationsUnreadWithAccounts = useCallback(
    async (notification: AtlasifyNotification) =>
      await markProductNotificationsUnread({ auth, settings }, notification),
    [auth, settings, markProductNotificationsUnread],
  );

  return (
    <AppContext.Provider
      value={{
        auth,
        isLoggedIn,
        loginWithAPIToken,
        logoutFromAccount,

        status,
        globalError,

        notifications,
        fetchNotifications: fetchNotificationsWithAccounts,

        markNotificationRead: markNotificationReadWithAccounts,
        markNotificationUnread: markNotificationUnreadWithAccounts,

        markProductNotificationsRead: markProductNotificationsReadWithAccounts,
        markProductNotificationsUnread:
          markProductNotificationsUnreadWithAccounts,

        settings,
        clearFilters,
        resetSettings,
        updateSetting,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
