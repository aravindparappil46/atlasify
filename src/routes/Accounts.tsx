import { PersonIcon as PersonIconOld } from '@primer/octicons-react';

import { type FC, useCallback, useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import Avatar from '@atlaskit/avatar';
import InviteTeamIcon from '@atlaskit/icon/glyph/invite-team';
import LockIcon from '@atlaskit/icon/glyph/lock';
import PersonIcon from '@atlaskit/icon/glyph/person';
import RefreshIcon from '@atlaskit/icon/glyph/refresh';
import SignOutIcon from '@atlaskit/icon/glyph/sign-out';
import StarIcon from '@atlaskit/icon/glyph/star';
import StarFilledIcon from '@atlaskit/icon/glyph/star-filled';
import Tooltip from '@atlaskit/tooltip';

import { Header } from '../components/Header';
import { AppContext } from '../context/App';
import { BUTTON_CLASS_NAME } from '../styles/atlasify';
import type { Account } from '../types';
import { getAccountUUID, refreshAccount } from '../utils/auth/utils';
import { cn } from '../utils/cn';
import { updateTrayIcon, updateTrayTitle } from '../utils/comms';
import { openAccountProfile } from '../utils/links';
import { saveState } from '../utils/storage';

export const AccountsRoute: FC = () => {
  const { auth, settings, logoutFromAccount } = useContext(AppContext);
  const navigate = useNavigate();

  const logoutAccount = useCallback(
    (account: Account) => {
      logoutFromAccount(account);
      navigate(-1);
      updateTrayIcon();
      updateTrayTitle();
    },
    [logoutFromAccount],
  );

  const setAsPrimaryAccount = useCallback((account: Account) => {
    auth.accounts = [account, ...auth.accounts.filter((a) => a !== account)];
    saveState({ auth, settings });
    navigate('/accounts', { replace: true });
  }, []);

  const login = useCallback(() => {
    return navigate('/login', { replace: true });
  }, []);

  return (
    <div className="flex h-screen flex-col" data-testid="accounts">
      <Header icon={PersonIconOld}>Accounts</Header>
      <div className="flex-grow overflow-x-auto px-8">
        <div className="mt-4 flex flex-col text-sm">
          {auth.accounts.map((account, i) => (
            <div
              key={getAccountUUID(account)}
              className="mb-4 flex items-center justify-between rounded-md bg-gray-100 p-2 dark:bg-gray-sidebar"
            >
              <div className="ml-2">
                <div>
                  <Tooltip content={account.user.name}>
                    <Avatar
                      name={account.user.name}
                      src={account.user.avatar}
                      size="medium"
                      onClick={() => openAccountProfile(account)}
                    />
                  </Tooltip>
                </div>
              </div>
              <div className="flex flex-col gap-1">
                <div className="text-sm font-semibold">{account.user.name}</div>
                <div className="text-xs">
                  <button
                    type="button"
                    className="flex flex-1 gap-1 cursor-pointer align-middle"
                    title="Open Host"
                  >
                    <PersonIcon label="Username" size="small" />
                    {account.user.login}
                  </button>
                </div>
                <div className="text-xs">
                  <button
                    type="button"
                    className="flex flex-1 gap-1 cursor-pointer align-middle"
                    title="Open Developer Settings"
                    // onClick={() => openDeveloperSettings(account)}
                  >
                    <LockIcon label="Authentication method" size="small" />
                    {account.method}
                  </button>
                </div>
              </div>

              <div>
                <button
                  type="button"
                  className={cn(BUTTON_CLASS_NAME, 'px-0', 'cursor-default')}
                  title="Primary account"
                  hidden={i !== 0}
                >
                  <StarFilledIcon
                    label="Primary account"
                    primaryColor="gold"
                    size="medium"
                  />
                </button>
                <button
                  type="button"
                  className={cn(BUTTON_CLASS_NAME, 'px-0')}
                  title="Set as primary account"
                  onClick={() => setAsPrimaryAccount(account)}
                  hidden={i === 0}
                >
                  <StarIcon label="Set as primary account" size="medium" />
                </button>
                <button
                  type="button"
                  className={cn(BUTTON_CLASS_NAME, 'px-0')}
                  title={`Refresh ${account.user.login}`}
                  onClick={async () => {
                    await refreshAccount(account);
                    navigate('/accounts', { replace: true });
                  }}
                >
                  <RefreshIcon
                    label={`Refresh ${account.user.login}`}
                    size="medium"
                  />
                </button>
                <button
                  type="button"
                  className={cn(BUTTON_CLASS_NAME, 'px-0')}
                  title={`Logout ${account.user.login}`}
                  onClick={() => logoutAccount(account)}
                >
                  <SignOutIcon
                    label={`Logout ${account.user.login}`}
                    size="medium"
                  />
                </button>
              </div>
            </div>
          ))}
        </div>
      </div>

      <div className="flex items-center justify-between bg-gray-200 px-8 py-1 text-sm dark:bg-gray-darker">
        <div className="font-semibold italic">Add new account</div>
        <div>
          <button
            type="button"
            className={BUTTON_CLASS_NAME}
            title="Add new account"
            onClick={login}
          >
            <InviteTeamIcon label="Add new account" size="medium" />
          </button>
        </div>
      </div>
    </div>
  );
};
