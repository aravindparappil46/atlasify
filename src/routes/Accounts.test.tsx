import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import {
  mockAtlassianCloudAccount,
  mockAuth,
  mockSettings,
} from '../__mocks__/state-mocks';
import { AppContext } from '../context/App';
import * as apiRequests from '../utils/api/request';
import * as comms from '../utils/comms';
import * as links from '../utils/links';

import { AccountsRoute } from './Accounts';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('routes/Accounts.tsx', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('General', () => {
    it('should render itself & its children', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      expect(screen.getByTestId('accounts')).toMatchSnapshot();
    });

    it('should go back by pressing the icon', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: mockAuth,
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByLabelText('Go Back'));
      expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
    });
  });

  describe('Account interactions', () => {
    it('open profile in external browser', async () => {
      const openAccountProfileMock = jest
        .spyOn(links, 'openAccountProfile')
        .mockImplementation();

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Profile'));

      expect(openAccountProfileMock).toHaveBeenCalledTimes(1);
      expect(openAccountProfileMock).toHaveBeenCalledWith(
        mockAtlassianCloudAccount,
      );
    });

    it('open host in external browser', async () => {
      const openExternalLinkMock = jest
        .spyOn(comms, 'openExternalLink')
        .mockImplementation();

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Host'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
      expect(openExternalLinkMock).toHaveBeenCalledWith('https://github.com');
    });

    it('open developer settings in external browser', async () => {
      const openExternalLinkMock = jest
        .spyOn(comms, 'openExternalLink')
        .mockImplementation();

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Open Developer Settings'));

      expect(openExternalLinkMock).toHaveBeenCalledTimes(1);
      expect(openExternalLinkMock).toHaveBeenCalledWith(
        'https://github.com/settings/tokens',
      );
    });

    it('should refresh account', async () => {
      const apiRequestAuthMock = jest.spyOn(apiRequests, 'apiRequestAuth');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Refresh octocat'));

      expect(apiRequestAuthMock).toHaveBeenCalledTimes(1);
      expect(apiRequestAuthMock).toHaveBeenCalledWith(
        'https://api.github.com/user',
        'GET',
        'token-123-456',
      );
      await waitFor(() =>
        expect(mockNavigate).toHaveBeenNthCalledWith(1, '/accounts', {
          replace: true,
        }),
      );
    });

    it('should logout', async () => {
      const logoutFromAccountMock = jest.fn();
      const updateTrayIconMock = jest.spyOn(comms, 'updateTrayIcon');
      const updateTrayTitleMock = jest.spyOn(comms, 'updateTrayTitle');

      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: {
                accounts: [mockAtlassianCloudAccount],
              },
              settings: mockSettings,
              logoutFromAccount: logoutFromAccountMock,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      fireEvent.click(screen.getByTitle('Logout octocat'));

      expect(logoutFromAccountMock).toHaveBeenCalledTimes(1);

      expect(updateTrayIconMock).toHaveBeenCalledTimes(1);
      expect(updateTrayIconMock).toHaveBeenCalledWith();
      expect(updateTrayTitleMock).toHaveBeenCalledTimes(1);
      expect(updateTrayTitleMock).toHaveBeenCalledWith();
      expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
    });
  });

  describe('Add new accounts', () => {
    it('should login with atlassian account', async () => {
      await act(async () => {
        render(
          <AppContext.Provider
            value={{
              auth: { accounts: [mockAtlassianCloudAccount] },
              settings: mockSettings,
            }}
          >
            <MemoryRouter>
              <AccountsRoute />
            </MemoryRouter>
          </AppContext.Provider>,
        );
      });

      expect(screen.getByTitle('Login with Atlassian App').hidden).toBe(false);

      fireEvent.click(screen.getByTitle('Login with Atlassian'));
      expect(mockNavigate).toHaveBeenNthCalledWith(1, '/login-api-token', {
        replace: true,
      });
    });
  });
});
