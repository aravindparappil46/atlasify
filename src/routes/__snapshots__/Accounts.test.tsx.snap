// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/Accounts.tsx General should render itself & its children 1`] = `
<div
  class="flex h-screen flex-col"
  data-testid="accounts"
>
  <div
    class="mx-8 mt-2 flex items-center justify-between py-2"
  >
    <button
      class="focus:outline-none"
      title="Go Back"
      type="button"
    >
      <svg
        aria-label="Go Back"
        class="hover:text-gray-400"
        fill="currentColor"
        focusable="false"
        height="20"
        role="img"
        style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
        viewBox="0 0 16 16"
        width="20"
      >
        <path
          d="M7.78 12.53a.75.75 0 0 1-1.06 0L2.47 8.28a.75.75 0 0 1 0-1.06l4.25-4.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042L4.81 7h7.44a.75.75 0 0 1 0 1.5H4.81l2.97 2.97a.75.75 0 0 1 0 1.06Z"
        />
      </svg>
    </button>
    <h3
      class="text-lg font-semibold flex items-center"
    >
      <legend
        class="mb-1 mt-2 font-semibold flex items-center"
        id="accounts"
      >
        <svg
          aria-hidden="true"
          class="mr-2"
          fill="currentColor"
          focusable="false"
          height="16"
          style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M10.561 8.073a6.005 6.005 0 0 1 3.432 5.142.75.75 0 1 1-1.498.07 4.5 4.5 0 0 0-8.99 0 .75.75 0 0 1-1.498-.07 6.004 6.004 0 0 1 3.431-5.142 3.999 3.999 0 1 1 5.123 0ZM10.5 5a2.5 2.5 0 1 0-5 0 2.5 2.5 0 0 0 5 0Z"
          />
        </svg>
        Accounts
      </legend>
    </h3>
  </div>
  <div
    class="flex-grow overflow-x-auto px-8"
  >
    <div
      class="mt-4 flex flex-col text-sm"
    >
      <div
        class="mb-4 flex items-center justify-between rounded-md bg-gray-100 p-2 dark:bg-gray-sidebar"
      >
        <div
          class="ml-2"
        >
          <div>
            <div
              role="presentation"
            >
              <div
                style="display: inline-block; position: relative; outline: 0;"
              >
                <button
                  aria-label="At Lassian"
                  class="css-1tnqmrq"
                  type="button"
                >
                  <img
                    alt="At Lassian"
                    class="css-13ep12v"
                    data-vc="avatar-image"
                    src="https://avatars.githubusercontent.com/u/583231?v=4"
                    style="border-radius: 50%;"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col gap-1"
        >
          <div
            class="text-sm font-semibold"
          >
            At Lassian
          </div>
          <div
            class="text-xs"
          >
            <button
              class="flex flex-1 gap-1 cursor-pointer align-middle"
              title="Open Host"
              type="button"
            >
              <span
                aria-label="Username"
                class="css-1wits42"
                data-vc="icon-undefined"
                role="img"
                style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
              >
                <svg
                  height="24"
                  role="presentation"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <g
                    fill="currentcolor"
                    fill-rule="evenodd"
                  >
                    <path
                      d="M6 14c0-1.105.902-2 2.009-2h7.982c1.11 0 2.009.894 2.009 2.006v4.44c0 3.405-12 3.405-12 0z"
                    />
                    <circle
                      cx="12"
                      cy="7"
                      r="4"
                    />
                  </g>
                </svg>
              </span>
              atlas
            </button>
          </div>
          <div
            class="text-xs"
          >
            <button
              class="flex flex-1 gap-1 cursor-pointer align-middle"
              title="Open Developer Settings"
              type="button"
            >
              <span
                aria-label="Authentication method"
                class="css-1wits42"
                data-vc="icon-undefined"
                role="img"
                style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
              >
                <svg
                  height="24"
                  role="presentation"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <g
                    fill="currentcolor"
                  >
                    <path
                      d="M16 11V9h-2V7.002A2 2 0 0 0 12 5c-1.102 0-2 .898-2 2.002V9H8v2H7v8h10v-8zm-2 0h-4V9h4zM8 9V7.002A4.004 4.004 0 0 1 12 3a4 4 0 0 1 4 4.002V9h.994A2.01 2.01 0 0 1 19 11.009v7.982c0 1.11-.897 2.009-2.006 2.009H7.006A2.01 2.01 0 0 1 5 18.991V11.01C5 9.899 5.897 9 7.006 9zm0 0h2v2H8zm6 0h2v2h-2z"
                    />
                    <circle
                      cx="12"
                      cy="15"
                      r="2"
                    />
                  </g>
                </svg>
              </span>
              API Token
            </button>
          </div>
        </div>
        <div>
          <button
            class="hover:text-gray-500 py-1 my-1 mx-2 focus:outline-none px-0 cursor-default"
            title="Primary account"
            type="button"
          >
            <span
              aria-label="Primary account"
              class="css-1afrefi"
              data-vc="icon-undefined"
              role="img"
              style="--icon-primary-color: gold; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
            >
              <svg
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="m12.072 17.284-3.905 2.053a1 1 0 0 1-1.451-1.054l.745-4.349-3.159-3.08a1 1 0 0 1 .554-1.705l4.366-.635 1.953-3.956a1 1 0 0 1 1.794 0l1.952 3.956 4.366.635a1 1 0 0 1 .555 1.705l-3.16 3.08.746 4.349a1 1 0 0 1-1.45 1.054z"
                  fill="currentcolor"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <button
            class="hover:text-gray-500 py-1 my-1 mx-2 focus:outline-none px-0"
            hidden=""
            title="Set as primary account"
            type="button"
          >
            <span
              aria-label="Set as primary account"
              class="css-1afrefi"
              data-vc="icon-undefined"
              role="img"
              style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
            >
              <svg
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="m12 16.373 3.98 2.193-.76-4.655 3.276-3.347-4.524-.69L12 5.687l-1.972 4.189-4.524.689L8.78 13.91l-.762 4.655zm0 2.283-3.016 1.662a2 2 0 0 1-2.939-2.075l.599-3.656-2.57-2.624a2 2 0 0 1 1.129-3.377l3.47-.528 1.518-3.224a2 2 0 0 1 3.618 0l1.519 3.224 3.47.528a2 2 0 0 1 1.127 3.377l-2.569 2.624.599 3.656a2 2 0 0 1-2.94 2.075z"
                  fill="currentcolor"
                />
              </svg>
            </span>
          </button>
          <button
            class="hover:text-gray-500 py-1 my-1 mx-2 focus:outline-none px-0"
            title="Refresh atlas"
            type="button"
          >
            <span
              aria-label="Refresh atlas"
              class="css-1afrefi"
              data-vc="icon-undefined"
              role="img"
              style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
            >
              <svg
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
              >
                <g
                  fill="currentcolor"
                  fill-rule="evenodd"
                >
                  <path
                    d="M8 6.003v2.995a1 1 0 1 0 2 0V5.102C10 4.494 9.507 4 8.9 4H5a1 1 0 0 0 0 2.003z"
                    fill-rule="nonzero"
                  />
                  <path
                    d="M9.428 18.018C7.351 16.989 6 14.807 6 12.37c0-2.266 1.167-4.319 3.02-5.425.48-.286.646-.922.371-1.421a.98.98 0 0 0-1.364-.386C5.557 6.611 4 9.35 4 12.37c0 3.248 1.802 6.158 4.57 7.529.498.247 1.095.026 1.332-.493.237-.52.025-1.141-.474-1.388"
                    fill-rule="nonzero"
                  />
                  <path
                    d="M14 15.002v3.896c0 .608.493 1.102 1.1 1.102H19a1 1 0 0 0 0-2.003h-3v-2.995a1 1 0 1 0-2 0"
                  />
                  <path
                    d="M14.097 4.596c-.237.52-.025 1.14.474 1.387 2.077 1.029 3.428 3.21 3.428 5.647 0 2.266-1.167 4.32-3.021 5.425a1.063 1.063 0 0 0-.37 1.42c.274.5.885.673 1.364.387 2.47-1.473 4.027-4.212 4.027-7.232 0-3.248-1.802-6.158-4.57-7.528A.96.96 0 0 0 15.002 4a1 1 0 0 0-.905.596"
                  />
                </g>
              </svg>
            </span>
          </button>
          <button
            class="hover:text-gray-500 py-1 my-1 mx-2 focus:outline-none px-0"
            title="Logout atlas"
            type="button"
          >
            <span
              aria-label="Logout atlas"
              class="css-1afrefi"
              data-vc="icon-undefined"
              role="img"
              style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
            >
              <svg
                height="24"
                role="presentation"
                viewBox="0 0 24 24"
                width="24"
              >
                <g
                  fill="currentcolor"
                  fill-rule="evenodd"
                >
                  <path
                    d="M4.977 11A.99.99 0 0 0 4 12c0 .551.437 1 .977 1h11.046A.99.99 0 0 0 17 12a.99.99 0 0 0-.977-1z"
                  />
                  <path
                    d="M6.23 8.31 3.3 11.27a1.05 1.05 0 0 0 0 1.48l2.93 2.96a1.03 1.03 0 0 0 1.47 0 1.05 1.05 0 0 0 0-1.48L5.5 12l2.2-2.22a1.05 1.05 0 0 0 0-1.48 1.03 1.03 0 0 0-1.47 0M15.5 3H12v2h7v14h-7v2h7.01c1.1 0 1.99-.89 1.99-1.99V5a1.99 1.99 0 0 0-1.99-2z"
                  />
                </g>
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="flex items-center justify-between bg-gray-200 px-8 py-1 text-sm dark:bg-gray-darker"
  >
    <div
      class="font-semibold italic"
    >
      Add new account
    </div>
    <div>
      <button
        class="hover:text-gray-500 py-1 px-2 my-1 mx-2 focus:outline-none"
        title="Add new account"
        type="button"
      >
        <span
          aria-label="Add new account"
          class="css-1afrefi"
          data-vc="icon-undefined"
          role="img"
          style="--icon-primary-color: currentColor; --icon-secondary-color: var(--ds-surface, #FFFFFF);"
        >
          <svg
            height="24"
            role="presentation"
            viewBox="0 0 24 24"
            width="24"
          >
            <g
              fill="currentcolor"
              fill-rule="evenodd"
            >
              <rect
                height="6"
                rx="1"
                width="2"
                x="18"
                y="5"
              />
              <rect
                height="2"
                rx="1"
                width="6"
                x="16"
                y="7"
              />
              <path
                d="M5 14c0-1.105.902-2 2.009-2h7.982c1.11 0 2.009.894 2.009 2.006v4.44c0 3.405-12 3.405-12 0z"
              />
              <circle
                cx="11"
                cy="7"
                r="4"
              />
            </g>
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
